### PR TITLE
Processing data in models

### PR DESCRIPTION
**Pull request by Abdulwhbhussain**
[Project requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md#project-requirements)
[Create models](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md#create-models)
Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png).
Set up associations between models.
Remember that author_id column in posts table should be the foreign_key for the users table.
[Use models to insert data](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md#use-models-to-insert-data)
Familiarize yourself with ['rails c' command](https://guides.rubyonrails.org/command_line.html#bin-rails-console).
Open rails console for your project.
Create at least one user by running the following code:
   first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
   second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')
Create at least 4 posts written by one of the users you created by running the following code:
   first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')
Create at least 6 posts comments for one of the posts you created by running the following code:
   Comment.create(post: first_post, user: second_user, text: 'Hi Tom!' )
Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.
[Create custom methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md#create-custom-methods)
Your models should include some additional methods.

Users
A method that returns the 3 most recent posts for a given user.
Posts
A method that updates the posts counter for a user.
A method which returns the 5 most recent comments for a given post.
Comments
A method that updates the comments counter for a post.
Likes
A method that updates the likes counter for a post.
Go to rails c and check if your methods are working.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  def recent_5_comments
+    comments.order(created_at: :desc).limit(5)
+  end
+
+  private
+
+  def increment_no_of_posts
+    author.update(posts_counter: author.posts_counter + 1)
+  end
```
Comment (Shedrack-Sunday) [inline app/models/post.rb]: - [ ] Your method uses `update` to manually increment the `posts_counter` attribute by fetching the current value, adding 1 to it, and then updating the database.
- It involves an additional database query to retrieve the current value of `posts_counter` before updating it. This is less efficient than using `increment!.`

- [ ] The `increment!` method is designed to safely update a counter or attribute in the database. It automatically generates an efficient SQL query to increment the specified column without the need to fetch the current value first. This makes it the preferred choice for incrementing counters.


### Sample code
```
  # app/models/post.rb
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/comment.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  belongs_to :user
+  belongs_to :post
+
+  after_create :increment_no_of_comments
+
+  private
+
+  def increment_no_of_comments
+    post.update(comments_counter: post.comments_counter + 1)
+  end
```
Comment (Shedrack-Sunday) [inline app/models/comment.rb]: - [ ] As I pointed out in the `Post` model, kindly use the same suggestion to use `increment!` built in method 👍

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  belongs_to :user
+  belongs_to :post
+
+  after_create :increment_no_of_likes
+
+  private
+
+  def increment_no_of_likes
+    post.update(likes_counter: post.likes_counter + 1)
+  end
```
Comment (Shedrack-Sunday) [inline app/models/like.rb]: - [x] Kindly make refrenece to my comments on the `Post` model regarding the use of `increment` 👍

**Review by Shedrack-Sunday — CHANGES_REQUESTED** [review]
Hello @Abdulwhbhussain   👋 

# CHANGES REQUIRED ♻️

Impressive work on the project so far👏
You've done extremely well so far in this project,  However, there are some adjustments you need to make to move to the next project. 💯



## Highlights   
- Good use of GitHub features ✔️   
- Nice work with Rails model✔️
- Kudus making the requirement asked      ✔️  

# - Required Changes  🌵  

- [ ] Kindly take note of the comments i made in the `models` to have a better looking code 👍  

# Optional suggestions 

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing really improves your application.



Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/models/like.rb (hunk 3 above)
Comment (Abdulwhbhussain) [reply app/models/like.rb]: Okay I got it! Let me Implement it.

File: app/models/post.rb (hunk 1 above)
Comment (Abdulwhbhussain) [reply app/models/post.rb]: Okay!

**Review by shella12 — APPROVED** [review]
Hi @Abdulwhbhussain, 


## Status : APPROVED 🥳 🥳 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- Used `increament!` with all requested methods ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**